### PR TITLE
Introduce new `Slider` shapes

### LINKED ISSUE #141842: Update `Slider` for Material 3 redesign
M3 sliders have an [updated look](https://m3.material.io/components/sliders/overview)

image
image

Since this occurred after the primary Material 3 transition, this could potentially be an option that starts as defaulting to `false`, switches to `true`, and then is eventually removed. Or a new widget as they are significantly different.

**Specification:** https://m3.material.io/components/sliders/specs
**December 2023 updates:** https://m3.material.io/components/sliders/overview#79de9466-a59c-4236-9099-735e67a38e53

### PR DESCRIPTION
**Pull request by TahaTesser**
fixes [Update `Slider` for Material 3 redesign](https://github.com/flutter/flutter/issues/141842)

### Description

This PR adds new `Slider` shapes `BarSliderThumbShape`, `ExpressiveRoundedRectSliderTrackShape`, and `RoundedRectSliderValueIndicatorShape` which are enabled when using the new `SliderThemeData.use2024SliderShapes` flag.

### Preview
 

| Thumb | Pressed Thumb |
| --------------- | --------------- |
|  |  |

### New shapes when dragged

| Light mode | Dark mode |
| --------------- | --------------- |
| <video src="https://github.com/flutter/flutter/assets/48603081/04edf145-0755-4978-9c65-2b125d71e525" /> |  <video src="https://github.com/flutter/flutter/assets/48603081/c6a5cc5c-2c42-4e40-a7b8-b9cd33fe03bd" /> |

### Stop indicator in the `RoundedRectSliderValueIndicatorShape` track shape
| Preview 1 | Preview 2 |
| --------------- | --------------- |
|  | <video src="https://github.com/flutter/flutter/assets/48603081/fc658d04-874e-4f5b-91ba-89f332c3ea14" /> |

|`trackGapSize` set to `0` | Overridden `trackGapSize` & `barThumbSize`  |
| --------------- | --------------- |
|  |  |

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Review by guidezpl — COMMENTED** [review]
Nice work, just a general comment, I think we should make it easier to use the new shapes, since they will become the new default. Rather than manually specifying them, developers could flip a flag on `ThemeData`.

**Comment by TahaTesser** [comment]
> Nice work, just a general comment, I think we should make it easier to use the new shapes, since they will become the new default. Rather than manually specifying them, developers could flip a flag on `ThemeData`.

Interesting idea, tho this is only specific to `Slider` (next `RangeSlider`). Maybe in the `SliderThemeData`? Do you've a name for this flag?

**Comment by guidezpl** [comment]
Perhaps `useUpdatedSliderShapes` or `use2024SliderShapes`? I think it would apply to all types of sliders

File: dev/tools/gen_defaults/lib/slider_template.dart
Comment (guidezpl) [inline dev/tools/gen_defaults/lib/slider_template.dart:97]: please add a TODO to use tokens

**Review by guidezpl — CHANGES_REQUESTED** [review]

**Comment by TahaTesser** [comment]
@guidezpl 
Appreciate the review! I'm about start traveling for Google I/O 2024. I'll address them after returning from states.

File: packages/flutter/lib/src/material/slider_theme.dart
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider_theme.dart:3690]: There are no terms used for the shapes in the guidelines. In Flutter, we come up with own Slider names.

Comment (guidezpl) [reply packages/flutter/lib/src/material/slider_theme.dart:3696]: How about `RoundedRectSliderTrackShape` or `RoundedRectM3SliderTrackShape`?

Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider_theme.dart:3696]: `RoundedRectSliderTrackShape` already exist, it is currently default track shape in both M3 and M2 in the Flutter slider.
I think it's best not to use M3 prefix in the shape names as each shape is unique and also backward supported to M2.

The new track shape supports a gap and has stop indicator. We need a name matching this.

Comment (guidezpl) [reply packages/flutter/lib/src/material/slider_theme.dart:3696]: True. `GappedSliderTrackShape` perhaps?

Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider_theme.dart:3696]: Perfection!

**Comment by TahaTesser** [comment]
Updated the PR to use new tokens, restored the current M3 defaults, and made minor improvements.

File: dev/tools/gen_defaults/lib/slider_template.dart
Comment (guidezpl) [inline dev/tools/gen_defaults/lib/slider_template.dart:46]: I think this token is available

File: packages/flutter/lib/src/material/slider.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/slider.dart:2032]: nit: please clarify that legacy = first version of the M3 slider and that the look has been updated this year

File: packages/flutter/lib/src/material/slider_theme.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/slider_theme.dart:617]: nit: please indicate that this flag is intended to migrate usages, after which it will be defaulted to true in a breaking change, and subsequently removed.

**Review by guidezpl — APPROVED** [review]
nice! looking for a second review

File: dev/tools/gen_defaults/lib/slider_template.dart
Comment (TahaTesser) [reply dev/tools/gen_defaults/lib/slider_template.dart:46]: I checked, it seems incorrect it uses secondary container color which makes inactive tick marks invisible on the inactive track.  I will file an issue as this lands.

File: packages/flutter/lib/src/material/slider_theme.dart
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider_theme.dart:617]: Updated

File: packages/flutter/lib/src/material/slider.dart
Comment (TahaTesser) [reply packages/flutter/lib/src/material/slider.dart]: Updated

**Comment by guidezpl** [comment]
Added an internal fix, Google testing will test again now. Please ping me if it's still not passing and I forget to check.

**Comment by TahaTesser** [comment]
> Added an internal fix, Google testing will test again now. Please ping me if it's still not passing and I forget to check.

Thanks so much, looks like it passes now! Waiting for the second review.

**Review by Piinks — COMMENTED** [review]
This LGTM overall, it's just the use2024SliderShapes that gave me pause. Is this something we expect folks to migrate to? What is the difference between M2/M3/M3-2024?

Could we add some docs that explain this a bit more? And if we want folks to migrate, we should have a migration guide to let them know to start moving. :)

**Comment by TahaTesser** [comment]
Pushed some doc updates.  

> What is the difference between M2/M3/M3-2024?

Great questions! It's all in the visuals. Instead of listing all the shapes, I'll add some `Slider` diagrams in the `[assets-for-api-docs](https://github.com/TahaTesser/assets-for-api-docs)` and mention them here.

>  we should have a migration guide to let them know to start moving. :)

Will write a migration for `flutter/website`

**Comment by Piinks** [comment]
Are we wanting folks to migrate so we can have an updated default shape? Will all of the shapes be available (backwards compatible) for folks to use after the migration (and removal of the use2024Shapes flag)? 
If both answers are yes, and it is a simple migration of setting the shape on the slider, it may be worth considering just updating the default and writing a migration guide rather than doing a multi-step flag migration.

**Comment by TahaTesser** [comment]
Yes, all shapes are backwards compatible. I love this idea, however changing the defaults will give different slider appearance and it would be overridden with regular SliderTheme. 

Shall I remove the use2024SliderShapes flag? If we update defaults and they can customise the slider then we don't need this flag.

**Comment by TahaTesser** [comment]
Discussed this with Kate offline, moving forward with default update and migration guide.

**Comment by TahaTesser** [comment]
While updating the defaults, I ran into an existing bug with `divisions` that could revert this new shapes update, it's a blocker.
This division bug can reproduced in the current Slider in both M2 and M3 configurations, I should fix this bug first in separate PR.

File: packages/flutter/lib/src/material/slider_theme.dart
Comment (navaronbracke) [inline packages/flutter/lib/src/material/slider_theme.dart:298]: What about `useMaterial3SliderShapes` ? I wonder if adding the year these were introduced in the specification is a good idea for the naming (or not)

Comment (Piinks) [reply packages/flutter/lib/src/material/slider_theme.dart:298]: M3 shapes were already introduced, this is a new set.

**Comment by TahaTesser** [comment]
https://github.com/flutter/flutter/pull/149594 just landed which unblocks this PR. 

I'll update this PR and writing migration guide

**Comment by TahaTesser** [comment]
Closing this PR for a new clean PR which will also be included in the migration guide.

**Pull request closed without merging**
